### PR TITLE
ScanAndSign context passing

### PR DESCRIPTION
**Pull request by agr**
Progress on https://github.com/NuGet/Engineering/issues/4482

Created a second version of `ScanAndSignMessageData` with a new `Context` property that reflect the change in `ScanAndSignMessage`. It is a simple key-value property bag that is passed as is to the scan/sign service.

Added code to fill that context with package id, version and the owner list. We already pass that information in different form to the service, so no changes in the amount of data passed, just some denormalization to not have the service understand NuGet packages in any depth.

### CODE REVIEW

**Review by drewgillies — APPROVED** [review]

**Review by lyndaidaii — DISMISSED** [review]

File: src/Validation.ScanAndSign.Core/IScanAndSignEnqueuer.cs [hunk 1]
```diff
@@ -14,15 +14,24 @@ public interface IScanAndSignEnqueuer
         /// </summary>
         /// <param name="validationId">Validation ID for which scan is requested.</param>
         /// <param name="nupkgUrl">Url of the package to scan.</param>
-        Task EnqueueScanAsync(Guid validationId, string nupkgUrl);
+        /// <param name="context">Request context to pass to the actual signing service.</param>
```
Comment (zhhyu) [inline src/Validation.ScanAndSign.Core/IScanAndSignEnqueuer.cs:17]: >signing [](http://example.com/codeflow?start=72&length=7)

The comment here is a little misleading for me.
Should it be scanning service for the first two APIs? The last two are scanning and signing service?

File: src/Validation.ScanAndSign.Core/ScanAndSignMessageSerializer.cs [hunk 2]
```diff
@@ -11,29 +11,49 @@ public class ScanAndSignMessageSerializer : IBrokeredMessageSerializer<ScanAndSi
...
+        private IBrokeredMessageSerializer<ScanAndSignMessageData2> _serializer2 =
+            new BrokeredMessageSerializer<ScanAndSignMessageData2>();
 
         public ScanAndSignMessage Deserialize(IBrokeredMessage message)
         {
-            var deserializedMessage = _serializer.Deserialize(message);
+            try
+            {
+                var deserializedMessage2 = _serializer2.Deserialize(message);
+                return new ScanAndSignMessage(
```
Comment (zhhyu) [inline src/Validation.ScanAndSign.Core/ScanAndSignMessageSerializer.cs:24]: nit: add an empty line above.

File: src/Validation.ScanAndSign.Core/ScanAndSignMessageSerializer.cs [hunk 3]
```diff
@@ -11,29 +11,49 @@ public class ScanAndSignMessageSerializer : IBrokeredMessageSerializer<ScanAndSi
...
+                    deserializedMessage2.OperationRequestType,
+                    deserializedMessage2.PackageValidationId,
+                    deserializedMessage2.BlobUri,
+                    deserializedMessage2.V3ServiceIndexUrl,
+                    deserializedMessage2.Owners,
+                    deserializedMessage2.Context);
+            }
+            catch (FormatException)
+            {
+                // do nothing, will try to deserialize with v1 below
```
Comment (zhhyu) [inline src/Validation.ScanAndSign.Core/ScanAndSignMessageSerializer.cs:34]: Will it be better to have a test case to cover the fallback behavior?

**Review by zhhyu — APPROVED** [review]

File: src/Validation.ScanAndSign.Core/ScanAndSignMessageSerializer.cs (hunk 3 above)
Comment (agr) [reply src/Validation.ScanAndSign.Core/ScanAndSignMessageSerializer.cs:35]: Added.

File: src/Validation.ScanAndSign.Core/IScanAndSignEnqueuer.cs (hunk 1 above)
Comment (agr) [reply src/Validation.ScanAndSign.Core/IScanAndSignEnqueuer.cs]: Updated the comment. There is a single service, and the context is irrelevant to the operation (scan or sign) requested.

**Review by zhhyu — APPROVED** [review]

**Pull request merged**